### DISCUSSION TITLE
posts/Go/operators/

### DISCUSSION THREAD

**Discussion by Q1mi** [Q1mi/BlogComments#364] [Announcements]
https://www.liwenzhou.com/posts/Go/03_operators/ 

 李文周的Blog

**Comment by cg430** [comment]
打卡

**Comment by chenk18783023351** [comment]
打卡

**Comment by heroliy** [comment]
666

**Comment by mytea003** [comment]
练习题呢，哈哈

**Comment by SmeagolYe** [comment]
done

**Comment by zhl6522** [comment]
var arr = [...]int{1,2,3,4,5,4,2,1,5,2,1}

var a22 = make([]int, len(arr), len(arr))

for _,v := range arr{

	a22[v]++;

}

for k,v := range a22 {

	if v == 1 {

		fmt.Println(k,v)

	}

}

**Comment by Q1mi** [comment]
> @zhl6522
> var arr = [...]int{1,2,3,4,5,4,2,1,5,2,1}
> 
> var a22 = make([]int, len(arr), len(arr))
> 
> for _,v := range arr{
> 
> 	a22[v]++;
> 
> }
> 
> for k,v := range a22 {
> 
> 	if v == 1 {
> 
> 		fmt.Println(k,v)
> 
> 	}
> 
> }

你的解法没问题，但是想一下我为什么要把这个练习题放在这里哦。^_^

**Comment by llystudy** [comment]
s := [5]int{1, 2, 3, 1, 2}

	fmt.Println(s[0] ^ s[1] ^ s[2] ^ s[3] ^ s[4])

**Comment by Treehuang** [comment]
```
nums := [...]int{1, 1, 2, 2, 3, 4, 4}
res := 0
for _, v := range nums {
    res ^= v
}
fmt.Println(res)
```

**Comment by LnkSaber** [comment]
func singleNumber(nums []int) int {
    var res  int = 0
    for _,v := range nums{
        res ^= v
    } 
    return res;
}


任何数和 00 做异或运算，结果仍然是原来的数，即 a ^ 0=aa⊕0=a。
任何数和其自身做异或运算，结果是 00，即 a ^ a=0a⊕a=0。
异或运算满足交换律和结合律，即 a ^ b ^ a=b ^ a ^ a=b ^ (a ^ a)=b ^ 0=ba⊕b⊕a=b⊕a⊕a=b⊕(a⊕a)=b⊕0=b。

**Comment by jacobwwang** [comment]
```
package main

import "fmt"

func main() {
	// 参与运算的两数各对应的二进位相异或，当两对应的二进位相异时，结果为1。
	// 相同，为 0
	nums := [...]int{1, 2, 3, 3, 2, 1, 6, 9, 8, 8, 6}

	n := nums[0]
	for i := 1; i < len(nums); i++ {
		n ^= nums[i]
	}

	fmt.Println(n)
}
```

**Comment by CoderYM** [comment]
有时候比较容易混淆什么时候用" :=" 什么时候用=

**Comment by Q1mi** [comment]
> @CoderYM
> 有时候比较容易混淆什么时候用" :=" 什么时候用= 

“:=”是声明变量并赋值的一种简写
"="就是赋值

**Comment by xjb714** [comment]
```
package main
import("fmt")
func main(){
	numar,num:=[]int{1,2,3,4,5,5,4,3,2},0
	for _,v:=range numar {num^=v}
	fmt.Println(num)//输出1
}

**Comment by brianchiang-tw** [comment]
有一堆数字，如果除了一个数字以外，其他数字都出现了两次，<br>
那么如何找到出现一次的数字？


```
package main

import(
	"fmt"
)


func fundUnique(array []int) int{

	// x XOR x outputs 0
	// use this property to cancel those numbers which appear twice
	// then only the unique element remains in xorSum

	xorSum := 0

	for _, number := range array{
		xorSum ^= number
	}
	return xorSum
}


func main(){

	test_array := []int{1,1,2,2,3,4,4,5,5,6,7,6,7}

	nuiqueElement := fundUnique( test_array )

	fmt.Printf("Unique element = %v", nuiqueElement)

	return
}
```

output: <br>
Unique element = 3

**Comment by linsanityHuang** [comment]
func singleNumber(nums []int) int {
    res := 0

    for _, n := range nums {
        res ^= n
    }

    return res
}

**Comment by windf17** [comment]
~~~go
func FindSingle(n []int) int {
	result := 0
	for _, i := range n {
		result ^= i
	}
	return result
}
~~~

**Comment by xiaogev5** [comment]
```go
package main
import "fmt"
func main()  {
	var arr = [...]int{1,2,2,4,4,5,5,5,6}
	var ifRepeat bool
	for i:=0;i<len(arr);i++{
		ifRepeat = true
		for j:=0;j<len(arr);j++{
			if i!=j && arr[j]==arr[i] {
				ifRepeat = false
			}
		}
		if ifRepeat {
			fmt.Println(arr[i],"只出现一次")
		}
	}
}
 ```

**Comment by QuYingying0628** [comment]
```
//统计数组中出现了1次的数字(方法1：每次遍历到一个数，都再去遍历一下这个数组看看有没有重复的数)
	var arr=[...]int{1,1,2,3,3,4,4,5,6,5,6}
	var flag bool
	for i:=0;i<len(arr);i++{
		flag=true
		for j:=0;j<len(arr);j++{
			//找到了和自己相同的数字
			if arr[i]==arr[j]&&i!=j{
				flag=false;
				break
			}
		}

		if flag{
			fmt.Println("只出现了1次的数字是",arr[i])
			break
		}
	}

	//统计数组中出现了1次的数字
	//(方法2：用累积异或的方式，因为相同数字异或后就是0，而0与任何数字异或就是那个数字本身，
	//所以累计异或后就可以得到唯一不重复的数字)
	result := 0
	for _,num:=range arr{
		result^=num
	}
	fmt.Println("只出现了1次的数字是",result)
```

**Comment by QuYingying0628** [comment]
> # 运算符
>
> - 如何定义一个数组？
>
>   - 推导方式下的：
>
>     - 先用var声明变量的名字
>
>     - 然后在等于赋值的时候先用[...]表示是一个数组类型
>
>     - 然后紧跟着写出数组中元素的类型，比如int
>
>     - 再然后就是数组的内容，内容是和java一样，用花括号去包裹的
>
>     - 样例：
>
>       - ```go
>         var arr=[...]int{1,1,2,3,3,4,4,5,6,5,6}
>         ```
>
> - 变量只是var了一下，然后在下面继续使用的时候是会报错的，没办法进行推导，也就是说推导方式定义变量的话，这个推导一定要和var的时候一起用
>
>   - 比如：
>     - 先var flag，然后在下面循环的时候写了一个flag=true，是会报错的
>     - 但是我写在一起，也就是var flag = true，这样就是正确的
>     - 或者在var的时候就把类型确定好，var flag bool，然后在for循环里面赋值是可以的
>   - 也就是说这个推导没有那么智能，只有和var写在一行才能推导，因为var的时候必须要确定下来这个类型变量所需要的空间大小
>
> - for循环书写的方式：
>
>   - 用下标去遍历数组的时候，int i=0是可以直接简化写成i:=0的
>
>   - 用foreach的方式去遍历的话：
>
>     - java中是item类型 item：集合
>
>     - go中是item:=range 集合的方式，理解起来就是定义了一个循环遍历item，然后item的循环值范围就是集合的数据内容，用range去表示这个范围
>
>     - 需要注意的是，这边添加了一个“—”匿名变量
>
>       - 疑问这个匿名变量是干嘛的，为什么会出现？
>
>     - 示例：
>
>       - ```go
>         for _, num := range arr {
>         ```
>
> - go中应该是没有对象的概念的，所以如何获取数组长度呢：
>
>   - 就是用len方法
>     - 疑问：len方法是哪里的？是包里面的吗？如果是包里面的，那为什么没有加包名？
>
> - fmt里面的Println的使用方法和java中的println不太一样
>
>   - java中是直接用加号做字符串的拼接
>   - go中是用逗号表示字符串的拼接
>
> - go中函数的定义方式：
>
>   - func表明我要定义的是一个函数
>
>   - 然后跟着函数名，后面括号写上参数
>
>   - 参数是如何写的呢？
>
>     - 因为传入的是参数，所以不用加var，var就是定义的时候用
>     - 还是老规矩，先写参数名，然后后面写类型
>     - 如果是数组类型，这个类型怎么写呢？
>       - 就是先写[]表明是数组，然后后面跟着数组中的元素类型，比如int
>
>   - 再后面就是函数的返回值类型
>
>     - 如果没有返回值类型的话，就直接不写就行
>
>   - 样例：
>
>     - ```go
>       func fundUnique(array []int) int
>       ```

****

**Comment by bluesbonewong** [comment]
```go
func findOnceNum(n int) int {
	res := n % 10
	n = n / 10

	for n > 0 {
		res ^= n % 10
		n = n / 10
	}

	return res
}
```

**Comment by xujiaxindudu** [comment]
> @LnkSaber
> func singleNumber(nums []int) int {
>     var res  int = 0
>     for _,v := range nums{
>         res ^= v
>     } 
>     return res;
> }
> 
> 
> 任何数和 00 做异或运算，结果仍然是原来的数，即 a ^ 0=aa⊕0=a。
> 任何数和其自身做异或运算，结果是 00，即 a ^ a=0a⊕a=0。
> 异或运算满足交换律和结合律，即 a ^ b ^ a=b ^ a ^ a=b ^ (a ^ a)=b ^ 0=ba⊕b⊕a=b⊕a⊕a=b⊕(a⊕a)=b⊕0=b。
> 
> 
> 

看了异或运算的方法，就好做了

**Comment by JimouChen** [comment]
```go
func testYiHuo(nums []int) int {
	var res = nums[0]
	for i := 1; i < len(nums); i++ {
		res ^= nums[i]
	}
	return res
}
```
- 🤔亦或解决

**Comment by Johnnylov** [comment]
package main
import("fmt")

func main(){
nums := [ ]int{2,3,4,5,4,3,2,}
res := 0
for _,n = range nums{
res ^= n
}
return res
}

遍历nums数组 用0与它进行异或
res ^ 2 = 2
res ^ 3 = 2 ^ 3
res ^ 4 = 2 ^ 3^ 4
res ^ 5 = (2 ^ 3^ 4) ^ 5
res ^ 4 = (2 ^ 3^ 4 ^ 5) ^ 4
res ^ 3 = (2 ^ 3^ 4 ^ 5^ 4) ^ 3
res^ 2 = (2 ^ 3 ^ 4^ 5 ^ 4^ 3) ^ 2
= (2 ^ 2) ^ (3 ^ 3)^  (4 ^ 4) ^ 5
= 0 ^ 0 ^0 ^ 5
= 5
